packages/core/contracts/bot-helpers/atomic-disputer/ReserveCurrencyDisputer.sol

```diff
@@ -0,0 +1,126 @@
+     * @param liquidationId index of the liquidation for the given sponsor.
+     * @param maxReserveTokenSpent maximum number of reserve tokens to spend in the trade. Bounds slippage.
+     * @param deadline abort the trade and dispute if the transaction is mined after this timestamp.
+     **/
+    function swapDispute(
+        address uniswapRouter,
+        address financialContract,
+        address reserveCurrency,
+        uint256 liquidationId,
+        address sponsor,
+        FixedPoint.Unsigned calldata maxReserveTokenSpent,
+        uint256 deadline
+    ) public {
+        IFinancialContract fc = IFinancialContract(financialContract);
+
+        // 1. Fetch information about the liquidation from the financial contract.
+        IFinancialContract.LiquidationData memory liquidationData = fc.liquidations(sponsor, liquidationId);
+
+        // 2. Compute the disputeBondAmount. Multiply by the unit collateral so the dispute bond is a percentage of the
+        // locked collateral after fees. To add fees we simply multiply the rawUnitCollateral by the cumulativeFeeMultiplier.
+        FixedPoint.Unsigned memory disputeBondAmount =
+            liquidationData.lockedCollateral.mul(fc.disputeBondPercentage()).mul(
+                (liquidationData.rawUnitCollateral).mul(fc.cumulativeFeeMultiplier())
+            );
+
+        // 3. Calculate required collateral. Cost of a dispute is the dispute bond + the final fee.
+        FixedPoint.Unsigned memory totalCollateralRequired = disputeBondAmount.add(liquidationData.finalFee);
+
+        // 4. Compute the collateral shortfall. This considers and collateral that is current in the contract.
+        FixedPoint.Unsigned memory collateralToBePurchased =
+            subOrZero(totalCollateralRequired, getCollateralBalance(fc));
+
+        // 5. If there is collateral to be purchased, buy it on uniswap with the reserve currency.
+        if (collateralToBePurchased.isGreaterThan(0) && reserveCurrency != fc.collateralCurrency()) {
+            IUniswapV2Router01 router = IUniswapV2Router01(uniswapRouter);
+            address[] memory path = new address[](2);
+            path[0] = reserveCurrency;
+            path[1] = fc.collateralCurrency();
+
+            TransferHelper.safeApprove(reserveCurrency, address(router), maxReserveTokenSpent.rawValue);
+            router.swapTokensForExactTokens(
+                collateralToBePurchased.rawValue,
+                maxReserveTokenSpent.rawValue,
+                path,
+                address(this),
+                deadline
+            );
+        }
+
+        // 6. Finally, submit the dispute.
+        TransferHelper.safeApprove(fc.collateralCurrency(), address(fc), totalCollateralRequired.rawValue);
+        fc.dispute(liquidationId, sponsor);
+    }
+
+    // Helper method to work around subtraction overflow in the case of: a - b with b > a.
+    function subOrZero(FixedPoint.Unsigned memory a, FixedPoint.Unsigned memory b)
```

Comment:
This seems like a method we should add to FixedPoint lib since we use it in many places now, wdyT?

Reply:
it's in the two range trader bots now :) totally could refactor this into Fixedpoint in the future.